File: frontend/tools/compute_coverage.dart
import 'dart:io';

void main(List<String> args) {
  final thresholdArg = args.firstWhere(
    (a) => a.startsWith('--threshold='),
    orElse: () => '--threshold=70',
  );
  final threshold = double.tryParse(thresholdArg.split('=')[1]) ?? 70.0;

  final f = File('coverage/lcov.info');
  if (!f.existsSync()) {
    stderr.writeln('coverage/lcov.info not found. Run `flutter test --coverage` first.');
    exit(2);
  }

  final lines = f.readAsLinesSync();

  final Map<String, List<String>> files = {};
  String? current;
  for (final line in lines) {
    if (line.startsWith('SF:')) {
      current = line.substring(3);
      files[current] = [];
    } else if (current != null) {
      files[current]!.add(line);
    }
  }

  int totalAll = 0;
  int coveredAll = 0;

  print('Per-file coverage:');
  for (final entry in files.entries) {
    final fname = entry.key;
    int total = 0;
    int covered = 0;
    for (final line in entry.value) {
      if (line.startsWith('DA:')) {
        final parts = line.substring(3).split(',');
        if (parts.length >= 2) {
          total++;
          final hits = int.tryParse(parts[1]) ?? 0;
          if (hits > 0) covered++;
        }
      }
    }
    if (total > 0) {
      final pct = covered / total * 100.0;
      print('${fname.replaceAll('\\\\', '/').split('/').takeLast(3).join('/')} : ${covered}/${total} = ${pct.toStringAsFixed(2)}%');
    }
    totalAll += total;
    coveredAll += covered;
  }

  if (totalAll == 0) {
    print('No DA lines found in lcov.info; coverage unknown.');
    exit(3);
  }

  final pctAll = coveredAll / totalAll * 100.0;
  final result = '${coveredAll}/${totalAll} = ${pctAll.toStringAsFixed(2)}%';
  print('\nOverall coverage: $result');

  if (pctAll < threshold) {
    stderr.writeln('Coverage $result is below threshold ${threshold.toStringAsFixed(2)}%');
    exit(4);
  }

  print('Coverage meets threshold ${threshold.toStringAsFixed(2)}%');
  exit(0);
}

extension<T> on Iterable<T> {
  Iterable<T> takeLast(int n) sync* {
    final list = toList();
    final start = list.length - n;
    for (var i = start < 0 ? 0 : start; i < list.length; i++) {
      yield list[i];
    }
  }
}

File: frontend/tools/compute_coverage.py
import sys
import os

def main():
    threshold_val = 70.0
    for arg in sys.argv:
        if arg.startswith('--threshold='):
            try:
                threshold_val = float(arg.split('=')[1])
            except ValueError:
                pass

    lcov_file = 'coverage/lcov.info'
    if not os.path.exists(lcov_file):
        print('coverage/lcov.info not found. Run `flutter test --coverage` first.', file=sys.stderr)
        sys.exit(2)

    with open(lcov_file, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    files = {}
    current = None
    for line in lines:
        line = line.strip()
        if line.startswith('SF:'):
            current = line[3:]
            files[current] = []
        elif current is not None:
            files[current].append(line)

    total_all = 0
    covered_all = 0

    print('Per-file coverage:')
    for fname, data in files.items():
        total = 0
        covered = 0
        for line in data:
            if line.startswith('DA:'):
                parts = line[3:].split(',')
                if len(parts) >= 2:
                    total += 1
                    try:
                        hits = int(parts[1])
                        if hits > 0:
                            covered += 1
                    except ValueError:
                        pass
        
        if total > 0:
            pct = covered / total * 100.0
            display_name = '/'.join(fname.replace('\\', '/').split('/')[-3:])
            print(f'{display_name} : {covered}/{total} = {pct:.2f}%')
        
        total_all += total
        covered_all += covered

    if total_all == 0:
        print('No DA lines found in lcov.info; coverage unknown.')
        sys.exit(3)

    pct_all = covered_all / total_all * 100.0
    result = f'{covered_all}/{total_all} = {pct_all:.2f}%'
    print(f'\nOverall coverage: {result}')

    if pct_all < threshold_val:
        print(f'Coverage {result} is below threshold {threshold_val:.2f}%', file=sys.stderr)
        sys.exit(4)
    
    print(f'Coverage meets threshold {threshold_val:.2f}%')
    sys.exit(0)

if __name__ == "__main__":
    main()
